### ISSUE TITLE
General feedback and read optimisation question

### ISSUE THREAD

**Issue by ut0mt8** [parquet-go/parquet-go#200]
Hello fellow maintainers,

So first of all thanks for providing this lib. It's always great to have free implementation of a format in GO.

This is not really an issue per say but a general feedbacks and some question.

First : as already mentioned by others I really struggled at first to make something as simple to read a whole parquet file.

`````

        schema := parquet.SchemaOf(&Counters{})
        r := parquet.NewGenericReader[Counters](file, schema)
        defer r.Close()

        i := 0
        for {
                parquetRows := make([]Counters, chunkSize)
                rr, err := r.Read(parquetRows)
                if err != nil {
                        if errors.Is(err, io.EOF) {
                                break
                        } else {
                                fmt.Printf("%s \n", err)
                                continue
                        }
                }

                i = i + rr
                if i%100000 == 0 {
                        fmt.Printf("%d %#v\n", i, parquetRows[0])
                        break
                }
         }
`````
There is a clear documentation issue. I think the main problems are there no source of truth and informations are spread among the Readme, the godoc (with some part outdated) , examples and even github issues... It should be simplified to everything in godoc with the Readme being the header. That's for the documentation.

After my second problem is Reading in an efficient way (let say fast).
So basically with the code above and my parquet file (ok 100 columns and snappy encoded) I had something like 45K row read by second (which is slow really slow). I was expecting something much much better and profiling a bit with pprof I'm see there's a lot of use of reflection (which will kill performance unsurprisingly).
btw I tested the another go parquet library which have even worse performance (sigh)

My scenario : I want to read all rows of my parquet files passing unmodified to another program. So I don't really need to have something super unmarshalled in a structure (giving the fact my structure is fixed). 

How I could go faster (avoiding reflection) ?

Thks

**Comment by achille-roussel** [comment]
Hello @ut0mt8, thanks for opening this issue.

First, on making reading entire files easier, these might be useful functions to look at:
- https://pkg.go.dev/github.com/parquet-go/parquet-go#Read
- https://pkg.go.dev/github.com/parquet-go/parquet-go#ReadFile

We always welcome contributions, so if you think there are updates to the README that would be useful to make, feel free to send a pull request and we can get that merged in 👍 

On the performance side, 45K rows/s, with 100 columns each, that's 4.5M values decoded per second, ~200ns per value. Considering the values had to be decompressed, decoded, and reconstructed into Go objects, it's not too bad. Now, can we make it faster?

If your application needs to reconstruct Go objects, there is going to be more work to do than if you can work with lower-level constructs. I would first suggest taking a look at [`parquet.Row`](https://pkg.go.dev/github.com/parquet-go/parquet-go#Row) and using the [`ReadRows`](https://pkg.go.dev/github.com/parquet-go/parquet-go#RowReader.ReadRows) method instead. This is a lower-level API that is much more efficient but can also be harder to use, so depending on the use case it may or may not be the right solution.

I would love to help improve performance of the read path, tho I'm going to need more material:
- Could you share a CPU and memory profile of your code highlighting the bottlenecks?
- Could you show the parquet schema of the files you are working with?

**Comment by ut0mt8** [comment]
Thanks for your response. Really appreciated. 
Will try to make a PR for the documentation part.

So for the functions you pointed me, unfortunately the parquet file could not fit entirely in memory ; or more precisely I cannot wait (or do not want to wait) for the whole read to happen cause it's too. My goal is to read and stream parquet (actually it's a loader to cassandra/scylla). 

So I guess I could have more luck with low level primitives. (but for now I don't understand how to use them)

Also getting to know a bit better parquet format I think we can at least read each RowGroup in // within a goroutine. 
Will share pprof but for now we are bounded at one CPU, and half of the time in github.com/parquet-go/parquet-go.(*Schema).Reconstruct. 

To me there is big room for improvements.

Funny or not I just code before the same loader from another format I had to decode (the sstable cassandra format) . I ended by writing the parser of this ### format. Performance was at least 4X and inspecting cpu profiling go should pass most of its time in system call nothing else.

**Comment by ut0mt8** [comment]
ah and the schema/info about my files

``````
File name: plof.parquet
Version: v1.0
Created By: parquet-mr version 1.12.3 (build aeb73eda8c39f452cc600c7d4bd98ac310c91743)
Num Rows: 8417527
Number of RowGroups: 4
Number of Real Columns: 82
Number of Columns: 82
Number of Selected Columns: 82
Column 0: device_id (BYTE_ARRAY/UTF8)
Column 1: os (BYTE_ARRAY/UTF8)
Column 2: key (BYTE_ARRAY/UTF8)
Column 3: click_7d (INT32)
Column 4: click_90d (INT32)
Column 5: imp_7d (INT32)
Column 6: imp_90d (INT32)
Column 7: inactivity_days (INT32)
Column 8: view_detail_3d (INT32)
Column 9: view_detail_60d (INT32)
Column 10: additional_1_3d (INT32)
Column 11: additional_1_60d (INT32)
Column 12: additional_2_3d (INT32)
Column 13: additional_2_60d (INT32)
Column 14: additional_3_3d (INT32)
Column 15: additional_3_60d (INT32)
Column 16: add_to_basket_3d (INT32)
Column 17: add_to_basket_60d (INT32)
Column 18: view_listing_3d (INT32)
Column 19: view_listing_60d (INT32)
Column 20: transaction_3d (INT32)
Column 21: transaction_60d (INT32)
Column 22: first_transaction_3d (INT32)
Column 23: first_transaction_60d (INT32)
Column 24: open_3d (INT32)
Column 25: open_60d (INT32)
Column 26: install_3d (INT32)
Column 27: install_60d (INT32)
Column 28: reattribution_3d (INT32)
Column 29: reattribution_60d (INT32)
Column 30: signup_3d (INT32)
Column 31: signup_60d (INT32)
Column 32: in_app_ad_3d (INT32)
Column 33: in_app_ad_60d (INT32)
Column 34: level_completed_3d (INT32)
Column 35: level_completed_60d (INT32)
Column 36: tutorial_completed_3d (INT32)
Column 37: tutorial_completed_60d (INT32)
Column 38: login_3d (INT32)
Column 39: login_60d (INT32)
Column 40: transaction_revenue_3d (DOUBLE)
Column 41: transaction_revenue_60d (DOUBLE)
Column 42: in_app_counters_data_source (BYTE_ARRAY/UTF8)
Column 43: view_detail_3d_discrete (INT32)
Column 44: view_detail_60d_discrete (INT32)
Column 45: additional_1_3d_discrete (INT32)
Column 46: additional_1_60d_discrete (INT32)
Column 47: additional_2_3d_discrete (INT32)
Column 48: additional_2_60d_discrete (INT32)
Column 49: additional_3_3d_discrete (INT32)
Column 50: additional_3_60d_discrete (INT32)
Column 51: add_to_basket_3d_discrete (INT32)
Column 52: add_to_basket_60d_discrete (INT32)
Column 53: view_listing_3d_discrete (INT32)
Column 54: view_listing_60d_discrete (INT32)
Column 55: transaction_3d_discrete (INT32)
Column 56: transaction_60d_discrete (INT32)
Column 57: first_transaction_3d_discrete (INT32)
Column 58: first_transaction_60d_discrete (INT32)
Column 59: open_3d_discrete (INT32)
Column 60: open_60d_discrete (INT32)
Column 61: install_3d_discrete (INT32)
Column 62: install_60d_discrete (INT32)
Column 63: reattribution_3d_discrete (INT32)
Column 64: reattribution_60d_discrete (INT32)
Column 65: signup_3d_discrete (INT32)
Column 66: signup_60d_discrete (INT32)
Column 67: in_app_ad_3d_discrete (INT32)
Column 68: in_app_ad_60d_discrete (INT32)
Column 69: level_completed_3d_discrete (INT32)
Column 70: level_completed_60d_discrete (INT32)
Column 71: tutorial_completed_3d_discrete (INT32)
Column 72: tutorial_completed_60d_discrete (INT32)
Column 73: login_3d_discrete (INT32)
Column 74: login_60d_discrete (INT32)
Column 75: transaction_revenue_3d_discrete (INT32)
Column 76: transaction_revenue_60d_discrete (INT32)
Column 77: user_revenue_per_action (DOUBLE)
Column 78: transaction_gaming_365d (INT32)
Column 79: transaction_casino_365d (INT32)
Column 80: transaction_puzzle_365d (INT32)
Column 81: transaction_365d (INT32)
--- Row Group: 0  ---
--- Total Bytes: 160656015  ---
--- Rows: 2610100  ---
Column 0
 Values: 2610100, Min: [48 48 48 48 48 57 56 55 50 53 52 99 48 52 48 48 52 48 56 54 52 50 56 52 97 98 101 48 48 101 48 98 102 57 101 51 51 98 101 101], Max: [102 102 102 102 102 99 100 101 53 53 51 101 97 51 54 51 100 98 51 53 99 98 101 52 54 52 55 99 56 99 101 50 52 53 50 97 49 48 55 55], Null Values: 0, Distinct Values: 0
 Compression: SNAPPY, Encodings: RLE PLAIN BIT_PACKED
 Uncompressed Size: 114849242, Compressed Size: 105550340
Column 1
 Values: 2610100, Min: [97], Max: [105], Null Values: 0, Distinct Values: 0
 Compression: SNAPPY, Encodings: RLE PLAIN_DICTIONARY BIT_PACKED
 Uncompressed Size: 409135, Compressed Size: 401746
Column 2
....
``````

**Comment by ut0mt8** [comment]
btw as expected reading RowGroups in // greatly enhanced throughput. If I tune my parquet writer to produce small rowGroup it should be even better.

**Comment by ut0mt8** [comment]
let's close this one. I ended with readRow group in // and manually check type . something like this :

`````
 // loop over row groups
        for _, rgrp := range pkt.ParquetFile.RowGroups() {

                reader := parquet.NewRowGroupReader(rgrp)
                defer reader.Close()

                for {
                        // read rows by chunkSize
                        rows := make([]parquet.Row, pkt.ChunkSize)

                        nrows, err := reader.ReadRows(rows)
                        if err != nil {
                                if errors.Is(err, io.EOF) {
                                        break
                                }
                        }

                        for r := 0; r < nrows; r++ {
                                // we know the number of columns
                                values := make([]any, len(pkt.Schema.Fields()))

                                for i, val := range rows[r] {
                                        // type conversion
                                        if val.IsNull() {
                                                values[i] = &gocql.UnsetValue
                                        } else if val.Kind() == parquet.ByteArray {
                                                values[i] = val.String()
                                        } else if val.Kind() == parquet.Int32 {
                                                values[i] = val.Int32()
                                        } else if val.Kind() == parquet.Double {
                                                values[i] = val.Double()
                                        }
                                }

                              
                        }
                }
        }

`````

it's fast enough

**Comment by achille-roussel** [comment]
Awesome, I'm glad you found a way to use the package to solve your problem!
